kafkabp/rack.go

```diff
@@ -251,11 +266,50 @@ var awsRackID = sync.OnceValues(func() (string, error) {
 // the same error will be logged at slog's warning level every time
 // AWSAvailabilityZoneRackID is called.
 func AWSAvailabilityZoneRackID() string {
-	id, err := awsRackID()
+	id, err := awsZoneRackID()
 	if err != nil {
 		awsRackFailure.Inc()
 		slog.Warn("Failed to get AWS availability zone as rack id", "err", err)
 		return ""
 	}
 	return id
 }
+
+// AWSAvailabilityZoneIDRackID is a RackIDFunc implementation that returns AWS
+// availability zone id as the rack id.
+//
+// It also caches the result globally, so if you have more than one
+// AWSAvailabilityZoneRackID in your process only the first one actually makes
+// the HTTP request, for example:
+//
+//	consumer1 := kafkabp.NewConsumer(kafkabp.ConsumerConfig{
+//	    RackID: kafkabp.AWSAvailabilityZoneIDRackID,
+//	    Topic:  "topic1",
+//	    // other configs
+//	})
+//	consumer2 := kafkabp.NewConsumer(kafkabp.ConsumerConfig{
+//	    RackID: kafkabp.AWSAvailabilityZoneIDRackID,
+//	    Topic:  "topic2",
+//	    // other configs
+//	})
+//
+// It uses AWS instance metadata HTTP API with 1second overall timeout and 1024
+// HTTP response read limits..
+//
+// If there was an error retrieving rack id through AWS instance metadata API,
+// the same error will be logged at slog's warning level every time
+// AWSAvailabilityZoneRackID is called.
+//
+// See [1] for differences between AWS availability zone and availability zone
+// id.
+//
+// [1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-availability-zones
+func AWSAvailabilityZoneIDRackID() string {
+	id, err := awsZoneIDRackID()
+	if err != nil {
+		awsRackFailure.Inc()
+		slog.Warn("Failed to get AWS availability zone id as rack id", "err", err)
```

Comment:
🔕 LogDNA will not show the error message to users when the error is moved into the structured portion of the log.  I typically recommend having the error as part of the human portion instead, since error is not really an interesting thing to filter against or aggregate.

Reply:
my experience is the opposite, that with errors in kv pair makes it easier to find exact errors (it's probably not that helpful if you want to search for partial error messages) on logdna, for use cases like "how many time did this error happen, did we have a spike during this time window?".

with a better logging ui (like GCP's log explorer) you can add an kv value to the summary line in the ui if needed, which is the best of both worlds, but we just don't have that unfortunately.